packages/js/product-editor/src/hooks/use-product-template/use-product-template.ts

```diff
@@ -16,13 +16,56 @@ declare global {
 	}
 }
 
+const matchesAllTemplateMetaFields = (
+	templateMeta: Metadata< string >[],
+	productMeta: Metadata< string >[]
+) =>
+	templateMeta.every( ( item ) =>
+		productMeta.find(
+			( productMetaEntry ) =>
+				productMetaEntry.key === item.key &&
+				productMetaEntry.value === item.value
+		)
+	);
+
+function templateDataMatchesProductData(
+	productTemplate: ProductTemplate,
+	product: Partial< Product >
+): boolean {
+	return Object.entries( productTemplate.productData ).every(
+		( [ key, value ] ) => {
+			if ( key === 'meta_data' ) {
+				return matchesAllTemplateMetaFields(
+					value,
+					product.meta_data || []
+				);
+			}
+
+			return product[ key ] === value;
+		}
+	);
+}
+
+function findBetterMatchTemplate( matchingTemplates: ProductTemplate[] ) {
+	return matchingTemplates.reduce(
+		( previous, current ) =>
+			Object.keys( current.productData ).length >
+			Object.keys( previous.productData ).length
+				? current
+				: previous,
+		matchingTemplates[ 0 ]
+	);
+}
+
 export const useProductTemplate = (
 	productTemplateId: string | undefined,
-	productType: ProductType | undefined
+	product: Partial< Product > | undefined
```

Comment:
Since we pass in `null` if the product doesn't exist, we should handle that in the signature.